subsys/bluetooth/mesh/access.c

```
@@ -2443,187 +2608,236 @@ static int read_comp_cb(const char *key, size_t len, settings_read_cb read_cb,
-	net_buf_simple_add_le16(&buf, total_size);
+	path = stored_page_path(type, page);
 
-	err = bt_mesh_metadata_get_page_0(&buf, 0);
-	if (err == -E2BIG) {
-		LOG_ERR("Metadata too large");
-		return err;
-	}
-	if (err) {
-		LOG_ERR("Failed to read models metadata: %d", err);
-		return err;
+	if (path == NULL) {
+		return 0;
 	}
 
-	LOG_DBG("bt/mesh/metadata len %d", buf.len);
-
-	err = settings_save_one("bt/mesh/metadata", buf.data, buf.len);
+	err = settings_load_subtree_direct(path, stored_page_size_cb, &size);
 	if (err) {
-		LOG_ERR("Failed to store models metadata: %d", err);
-	} else {
-		LOG_DBG("Stored models metadata");
+		LOG_ERR("Failed getting stored page size for %sdata page %d: %d",
+			type == PAGE_TYPE_COMP ? "comp " : "meta", page, err);
+		return 0;
 	}
 
-	return err;
+	return size;
 }
+#endif /* CONFIG_BT_MESH_HIGH_DATA_PAGES */
 
-int bt_mesh_models_metadata_read(struct net_buf_simple *buf, size_t offset)
+static size_t page_size_get(enum page_type type, uint8_t page)
 {
-	NET_BUF_SIMPLE_DEFINE(stored_buf, CONFIG_BT_MESH_MODELS_METADATA_PAGE_LEN);
-	size_t original_len = buf->len;
-	int err;
+#ifdef CONFIG_BT_MESH_HIGH_DATA_PAGES
+	size_t size;
 
-	if (!IS_ENABLED(CONFIG_BT_SETTINGS)) {
-		return -ENOTSUP;
+	if (page >= 128) {
+		size = stored_page_size_get(type, page);
+		if (size == 1) {
+			return 0;
+		}
+		if (size > 1) {
+			return size;
+		}
 	}
+#endif
+	return current_page_size(type, page % 128);
+}
 
-	net_buf_simple_init(&stored_buf, 0);
+static int get_page_contents(struct net_buf_simple *buf, enum page_type type, uint8_t page,
+			     size_t offset, bool allow_partial_elems)
+{
+#ifdef CONFIG_BT_MESH_HIGH_DATA_PAGES
+	int err;
 
-	err = settings_load_subtree_direct("bt/mesh/metadata", read_comp_cb, &stored_buf);
-	if (err) {
-		LOG_ERR("Failed reading models metadata: %d", err);
-		return err;
+	if (page >= 128) {
+		err = stored_page_read(buf, type, page, offset, allow_partial_elems);
+		if (err != -ENOENT) {
```

What if err == 0? current_page_contents will only be called, if err == -ENOENT. Is that correct?

---

Yes, see my explanation for similar comment in `page_size_get`

---

Oh, no, wait, this is a different question: Yes, I decided to do it this way, because if err != -ENOENT, then we don't know if there actually is some data in settings but we could not read it. In that case, it would be wrong to return the current page in my opinion, since there could be a problem with settings read but the application has _actually_ set the page data, we just couldn't read it. Then it's better to return an error than to assume that there is no data and fall back to current contents.

---

Re-reading this I thought to explain the err == 0 case: if err == 0, we succesfully found contents in settings and wrote them to `buf`, so then we want to return this 0. The only reason we want to proceed to `current_page_contents` is if there is no page data in flash (i.e. if it returns ENOENT).

---

Ok, a comment in the code would be nice. I will not remember that next time I read the code.